design-proposals/shadow-node.md

```diff
@@ -0,0 +1,352 @@
+  | for label in `newNonKubevirtLabels`: {(label in `oldNonKubevirtLabels`) AND `newLabels`[label] == `oldLabels`[label]}             | virt-handler user cannot update non kubevirt-owned labels                      |                                                                                                                                                                        |
+
+### Backwards compatibility:
+validationAdmissionPolicy [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it should be OK for deployment on Kubevirt main and next minor release branches.
+However, since kubevirt needs to be supported up to 2 k8s versions back, then the deployment of these new objects needs to be conditioned by the k8s version deployed. 
+The virt-operator deployment logic should be: 
+```
+if k8s-ver >= v1.30 
+then
+  deploy the objects
+else
+  deploy, but only log on failure
+```
+Once k8s v1.29 and below is not supported on kubevirt, this logic should be removed.
+
+### Next step: Prevent patches from other nodes' virt-handlers
+It is possible to expand the rules of the validationAdmissionPolicy to also prohibit virt-handler running on a node from patching other nodes.
+This could be defined by setting a new variable `requestHasNodeName` which will be true if request has `userInfo.extra.authentication.kubernetes.io/node-name` field, then adding this validation:
+```
+!`requestHasNodeName` OR (object.Name == request.nodeName)
+```
+However, this field is [beta](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#additional-metadata-in-pod-bound-tokens) on k8s 1.30  and is currently protected by the [ServiceAccountTokenPodNodeInfo](https://github.com/kubernetes/kubernetes/blob/c4bce63d9886e5f1fc00f8c3b5a13ea0d2bdf772/pkg/features/kube_features.go#L753) k8s feature gate.
+This field can be added once it is GA.
+
+Pros:
+* Avoid adding CRDs or changing the current virt-handler code to fix the CVE issue.
+* A k8s built-in feature which is straightforward to maintain and configure.
+* validationAdmissionPolicy is [going GA](https://kubernetes.io/blog/2024/03/12/kubernetes-1-30-upcoming-changes/) on k8s 1.30, so it could be used for the next Kubevirt minor release.
+
+Cons:
+* Currently, can't add the next step that prevents patches from other nodes' virt-handlers.
+
+## Solution option #2: ShadowNode CRD
+The security issue is being addressed by introducing a new light weight CRD: "ShadowNode" for each Kubernetes node.
+Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.
```

Comment:
Actually, I'm considering whether this solution could be viable, particularly if it's used alongside the [scheduling readiness feature of k8s](https://kubernetes.io/blog/2022/12/26/pod-scheduling-readiness-alpha/), which is expected to be [GAed in k8s 1.30](https://github.com/kubernetes/enhancements/issues/3521#issuecomment-1843817972). What are your thoughts on this approach?

The flow I've been thinking about goes like this:

1. virt-handler is deployed and sets the `ShadowNode` Objects with the proper annotations and labels, just like you mentioned.
2. Whenever a vmi is created, virt-api will mutate the virt-launcher pod with a dedicated SchedulingGate.
3. virt-controller will detect virt-launcher pods with the SchedulingGate and do the following for each launcher pod:
    * Calculate suitable nodes for scheduling from the kubevirt perspective, let's mark it as \<nodeList\>.
    * Remove the SchedulingGate from the launcher pod and add node affinity to allow scheduling only to nodes in \<nodeList\>.

By leveraging scheduling readiness, we might be able to avoid node modifications entirely within KubeVirt, as suggested in [this comment](https://github.com/kubevirt/community/pull/246/files#r1586066156), and will also allow us to remove all the boilerplate and implementation details of KubeVirt from the nodes.  Additionally, it might address concerns about users with node access gaining visibility into hardware information. 
Any thoughts?

FYI @vladikr

Reply:
@Barakmor1 I think this is a very interesting idea! I'm not 100% sure we need a gate. 
Virt-controller is the one that creates the pods, so it can calculate the list of suitable nodes before posting the Pod - and do the same during migration.
One possible issue is (as @xpivarc pointed out offline) that this list may become extensive. 

@rmohr FYI

Reply:
> @Barakmor1 I think this is a very interesting idea! I'm not 100% sure we need a gate.
Virt-controller is the one that creates the pods, so it can calculate the list of suitable nodes before posting the Pod - and do the same during migration.

 but if no node match a launcher it will remain pending forever as the <nodeList> will remain empty.

> One possible issue is (as @xpivarc pointed out offline) that this list may become extensive.

I think that [the maximum amount of nodes is 5000](https://kubernetes.io/docs/setup/best-practices/cluster-large/)

Reply:
My 2 cents - +1 to all efforts to use alternative ways to avoid the need to patch the node. 
If we could find good alternatives to all the places virt-handler is patching the node - then we could remove the patch permission entirely.
But even if we remove the node-labeller component we  currently still have other places where we patch the node (for example heartbeat) and until we remove them all - the need to restrict the node patches is still relevant.

Reply:
> But even if we remove the node-labeller component we currently still have other places where we patch the node (for example heartbeat) and until we remove them all - the need to restrict the node patches is still relevant.

Can't we patch the `shadowNode` with heartbeat labels as well?

Reply:
> > But even if we remove the node-labeller component we currently still have other places where we patch the node (for example heartbeat) and until we remove them all - the need to restrict the node patches is still relevant.
> 
> Can't we patch the `shadowNode` with heartbeat labels as well?

We can, but I believe we can solve it in a simpler, more k8s approach. Will issue a proposal about it in the near future.